File: submissions/dustboy/examples/verifier.py
"""ArraMQ v2 — per-message E2E verifier (Layer 2, the core).

Broker-agnostic: a subscriber, or a verifying bridge that republishes only valid
messages, calls Verifier.verify() on each incoming MQTT message. Closes the three
properties the workshop cohort each had only partially:
  1. topic-in-signed-body    EIP-712 Publish.topic must equal the DELIVERY topic
  2. real EIP-712            chainId is in the domain -> enters the digest
  3. persisted monotonic seq Redis Lua CAS -> survives verifier restart/scale AND
                             is atomic (no get-then-put TOCTOU race)

MQTT payload wire format (JSON):
  { "msg": {"topic","payloadHash","issuedAt","seq"}, "data": <str>, "sig": "0x.." }
Publisher address is RECOVERED from sig, never trusted from the wire.

  pip install eth-account eth-utils redis
  # self-test needs a redis:  docker compose up -d redis  then  python verifier.py
"""
import json, os, time
from eth_account import Account
from eth_account.messages import encode_typed_data
from eth_utils import keccak
import redis

CHAIN_ID = 20260619
MAX_AGE  = 600          # connect/message freshness window (s)
SKEW     = 60           # future-dated tolerance (s)

DOMAIN = {"name": "ArraMQ", "version": "1", "chainId": CHAIN_ID}
TYPES = {
    "EIP712Domain": [
        {"name": "name", "type": "string"},
        {"name": "version", "type": "string"},
        {"name": "chainId", "type": "uint256"},
    ],
    "Publish": [
        {"name": "topic", "type": "string"},
        {"name": "payloadHash", "type": "bytes32"},
        {"name": "issuedAt", "type": "uint256"},
        {"name": "seq", "type": "uint256"},
    ],
}

# atomic monotonic compare-and-set — no get-then-put race (eventual-consistency safe)
_CAS = """
local cur = tonumber(redis.call('GET', KEYS[1]) or '-1')
if tonumber(ARGV[1]) > cur then redis.call('SET', KEYS[1], ARGV[1]); return 1 else return 0 end
"""


class Verifier:
    def __init__(self, redis_url=None, allowlist=None):
        self.r = redis.Redis.from_url(redis_url or os.environ.get("REDIS_URL", "redis://localhost:6379/0"))
        self._cas = self.r.register_script(_CAS)
        self.allowlist = {a.lower() for a in allowlist} if allowlist else None

    def verify(self, delivery_topic, raw):
        """Returns (ok: bool, info: signer-address-or-reason)."""
        try:
            env = json.loads(raw)
            m = env["msg"]
            full = {"types": TYPES, "primaryType": "Publish", "domain": DOMAIN, "message": {
                "topic": m["topic"],
                "payloadHash": bytes.fromhex(m["payloadHash"][2:]),
                "issuedAt": int(m["issuedAt"]),
                "seq": int(m["seq"]),
            }}
            signer = Account.recover_message(encode_typed_data(full_message=full),
                                             signature=env["sig"]).lower()
        except Exception as e:
            return False, f"bad-signature:{e}"

        # 0. IDENTITY: the topic namespace must belong to the recovered signer.
        #    ArraMQ ACL is user/<addr>/# -> a publish to user/0xAAA/.. must recover to 0xAAA.
        #    This also makes a wrong-chainId signature fail: a different domain recovers a
        #    DIFFERENT address, which then won't own the topic. (allowlist for shared topics.)
        parts = delivery_topic.split("/")
        if len(parts) >= 2 and parts[0] == "user":
            if parts[1].lower() != signer:
                return False, "topic-not-owned-by-signer"
        elif self.allowlist is not None and signer not in self.allowlist:
            return False, "not-authorized"
        if m["topic"] != delivery_topic:                                   # 1. topic-binding
            return False, "topic-mismatch"
        if "0x" + keccak(env["data"].encode()).hex() != m["payloadHash"].lower():  # 2. payload-binding
            return False, "payload-hash-mismatch"
        now, ia = int(time.time()), int(m["issuedAt"])                     # 3. freshness
        if now - ia > MAX_AGE or ia - now > SKEW:
            return False, "stale-or-future"
        if not self._cas(keys=[f"seq:{signer}:{delivery_topic}"], args=[int(m["seq"])]):  # 4. monotonic
            return False, "replay-or-stale-seq"
        return True, signer


# --- runnable self-test (proves the 3 properties) ---
if __name__ == "__main__":
    acct = Account.from_key("0x" + "11" * 32)
    addr = acct.address

    def sign_pub(topic, data, seq, ia=None, chain=CHAIN_ID):
        ia = ia or int(time.time())
        dom = {"name": "ArraMQ", "version": "1", "chainId": chain}
        full = {"types": TYPES, "primaryType": "Publish", "domain": dom, "message": {
            "topic": topic, "payloadHash": bytes.fromhex(keccak(data.encode()).hex()),
            "issuedAt": ia, "seq": seq}}
        sig = acct.sign_message(encode_typed_data(full_message=full)).signature.hex()
        return json.dumps({"msg": {"topic": topic, "payloadHash": "0x" + keccak(data.encode()).hex(),
                                   "issuedAt": ia, "seq": seq}, "data": data, "sig": sig})

    v = Verifier()
    T = f"user/{addr.lower()}/telemetry"        # ArraMQ namespace owned by the signer
    CMD = f"user/{addr.lower()}/cmd"
    v.r.delete(f"seq:{addr.lower()}:{T}")
    cases = [
        ("valid",                v.verify(T, sign_pub(T, "pm25=12", 1)),               True),
        ("replay same seq",      v.verify(T, sign_pub(T, "pm25=12", 1)),               False),
        ("monotonic next",       v.verify(T, sign_pub(T, "pm25=13", 2)),               True),
        ("tampered data",        v.verify(T, sign_pub(T, "pm25=13", 3).replace("pm25=13", "pm25=99")), False),
        ("wrong delivery topic", v.verify(CMD, sign_pub(T, "open", 4)),                False),
        ("wrong chainId",        v.verify(T, sign_pub(T, "pm25=14", 5, chain=1)),      False),
        ("foreign namespace",    v.verify("user/0xdeadbeef/telemetry", sign_pub("user/0xdeadbeef/telemetry", "x", 6)), False),
    ]
    ok = True
    for name, (got, info), want in cases:
        flag = "PASS" if got == want else "FAIL"
        if got != want: ok = False
        print(f"[{flag}] {name:22} -> ok={got} ({info})")
    print("ALL PASS" if ok else "SELF-TEST FAILED"); raise SystemExit(0 if ok else 1)

File: submissions/vessel/examples/verifier.ts
/**
 * ArraMQ — Message Verifier (broker-agnostic)
 * Vessel proposal: topic-binding + payload-binding + Redis CAS seq
 *
 * Subscribes to fleet/+/# and verifies every message.
 * Also includes a self-test (no broker, no Redis required).
 *
 * Usage:
 *   MQTT_URL=mqtt://localhost:1883 REDIS_URL=redis://localhost:6379 bun run verifier.ts
 *   bun run verifier.ts --self-test
 */

import { recoverTypedDataAddress, keccak256, toBytes, getAddress } from "viem";
import { privateKeyToAccount, signTypedData } from "viem/accounts";
import { createClient } from "redis";
import mqtt from "mqtt";

const DOMAIN = {
  name: "ARRA-MQTT",
  version: "1",
  chainId: 20260619,
} as const;

const PUBLISH_TYPES = {
  Publish: [
    { name: "topic",       type: "string"  },
    { name: "payloadHash", type: "bytes32" },
    { name: "issuedAt",    type: "uint256" },
    { name: "seq",         type: "uint256" },
  ],
} as const;

const MAX_AGE_MS = 300_000;
const CLOCK_SKEW_MS = 5_000;

// Atomic Lua CAS — reject if seq not strictly increasing
const CAS_SCRIPT = `
local key = KEYS[1]
local incoming = tonumber(ARGV[1])
local last = tonumber(redis.call("GET", key) or "0")
if incoming > last then
  redis.call("SET", key, incoming)
  return 1
else
  return 0
end
`;

interface VerifyResult {
  ok: boolean;
  signer?: string;
  reason?: string;
}

async function verifyMessage(
  deliveryTopic: string,
  rawPayload: string,
  redis: Awaited<ReturnType<typeof createClient>> | null,
): Promise<VerifyResult> {
  let parsed: any;
  try {
    parsed = JSON.parse(rawPayload);
  } catch {
    return { ok: false, reason: "invalid JSON" };
  }

  const { data, payloadHash, issuedAt, seq, sig, signer } = parsed;
  if (!data || !payloadHash || !issuedAt || !seq || !sig) {
    return { ok: false, reason: "missing fields" };
  }

  // 1. Recover signer
  let recovered: string;
  try {
    recovered = await recoverTypedDataAddress({
      domain: DOMAIN,
      types: PUBLISH_TYPES,
      primaryType: "Publish",
      message: {
        topic: deliveryTopic,
        payloadHash: payloadHash as `0x${string}`,
        issuedAt: BigInt(issuedAt),
        seq: BigInt(seq),
      },
      signature: sig as `0x${string}`,
    });
  } catch {
    return { ok: false, reason: "sig recovery failed" };
  }

  // 2. Topic-binding (signed.topic must equal delivery topic)
  // (already enforced: we pass deliveryTopic into the recovery above)
  // If broker rerouted, recovery would succeed but the signer field mismatch catches tampering
  if (signer && getAddress(signer) !== getAddress(recovered)) {
    return { ok: false, reason: "signer field mismatch (possible replay on different identity)" };
  }

  // 3. Payload-binding
  const expectedHash = keccak256(toBytes(data));
  if (payloadHash !== expectedHash) {
    return { ok: false, reason: "payload hash mismatch (tampered)" };
  }

  // 4. Freshness
  const age = Date.now() - Number(issuedAt);
  if (age < -CLOCK_SKEW_MS || age > MAX_AGE_MS) {
    return { ok: false, reason: `stale message, age=${age}ms` };
  }

  // 5. Monotonic seq (Redis CAS)
  if (redis) {
    const seqKey = `seq:${recovered}:${deliveryTopic}`;
    const result = await redis.eval(CAS_SCRIPT, { keys: [seqKey], arguments: [seq.toString()] });
    if (result !== 1) {
      return { ok: false, reason: `replay: seq ${seq} not > last` };
    }
  }

  return { ok: true, signer: recovered };
}

// Self-test (no broker/Redis)
async function selfTest() {
  const TEST_PK = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80" as `0x${string}`;
  const acct = privateKeyToAccount(TEST_PK);
  const topic = "fleet/0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266/test";

  async function makeMsg(overrides: Record<string, any> = {}) {
    const issuedAt = BigInt(Date.now());
    const seq = BigInt(overrides.seq ?? 1);
    const data = overrides.data ?? "hello fleet";
    const payloadHash = keccak256(toBytes(data)) as `0x${string}`;
    const signTopic = overrides.signTopic ?? topic;

    const sig = await signTypedData({
      privateKey: TEST_PK,
      domain: DOMAIN,
      types: PUBLISH_TYPES,
      primaryType: "Publish",
      message: { topic: signTopic, payloadHash, issuedAt, seq },
    });

    return JSON.stringify({ data: overrides.data ?? data, payloadHash, issuedAt: issuedAt.toString(), seq: seq.toString(), sig, signer: acct.address });
  }

  // In-memory seq store for self-test
  const seqStore = new Map<string, bigint>();
  const mockRedis = {
    eval: async (_: string, opts: { keys: string[]; arguments: string[] }) => {
      const key = opts.keys[0];
      const incoming = BigInt(opts.arguments[0]);
      const last = seqStore.get(key) ?? 0n;
      if (incoming > last) { seqStore.set(key, incoming); return 1; }
      return 0;
    },
  } as any;

  const cases: Array<[string, () => Promise<[string, string]>, boolean]> = [
    ["valid", async () => [topic, await makeMsg()], true],
    ["replay-same-seq", async () => [topic, await makeMsg({ seq: 1 })], false],
    ["monotonic-next-seq", async () => [topic, await makeMsg({ seq: 2 })], true],
    ["tampered-data", async () => {
      const raw = JSON.parse(await makeMsg({ seq: 3 }));
      raw.data = "tampered!";
      return [topic, JSON.stringify(raw)];
    }, false],
    ["wrong-delivery-topic", async () => [topic + "/wrong", await makeMsg({ seq: 3, signTopic: topic + "/wrong" })], false],
  ];

  console.log("\n[self-test] ArraMQ verifier (Vessel)");
  let pass = 0, fail = 0;
  for (const [name, build, expected] of cases) {
    const [t, payload] = await build();
    const result = await verifyMessage(t, payload, mockRedis);
    const got = result.ok;
    const ok = got === expected;
    console.log(`  ${ok ? "✓" : "✗"} ${name}: ok=${got}${result.reason ? ` (${result.reason})` : ""}`);
    ok ? pass++ : fail++;
  }
  console.log(`\n  ${pass}/${pass + fail} passed`);
}

async function main() {
  if (process.argv.includes("--self-test")) {
    await selfTest();
    return;
  }

  const redis = createClient({ url: process.env.REDIS_URL ?? "redis://localhost:6379" });
  await redis.connect();

  const client = mqtt.connect(process.env.MQTT_URL ?? "mqtt://localhost:1883");

  client.on("connect", () => {
    console.log("[verifier] connected, subscribing fleet/+/#");
    client.subscribe("fleet/+/#", { qos: 1 });
  });

  client.on("message", async (deliveryTopic, buffer) => {
    const result = await verifyMessage(deliveryTopic, buffer.toString(), redis);
    if (result.ok) {
      console.log(`[✓] ${deliveryTopic} signer=${result.signer}`);
    } else {
      console.log(`[✗] ${deliveryTopic} REJECTED: ${result.reason}`);
    }
  });
}

main().catch(console.error);
